src/goto-symex/goto_symex_state.cpp

```diff
@@ -538,128 +611,170 @@ bool goto_symex_statet::l2_thread_write_encoding(
-      expr.type() = to_array_type(index_expr.array().type()).subtype();
+    auto renamed_array = rename_address<level>(index_expr.array(), ns);
 
-      // the index is not an address
-      index_expr.index() = rename<level>(std::move(index_expr.index()), ns);
-    }
-    else if(expr.id()==ID_if)
+    // the index is not an address
+    auto renamed_index = rename_expr<level>(index_expr.index(), ns);
+
+    if(renamed_array.has_value() || renamed_index.has_value())
     {
-      // the condition is not an address
-      if_exprt &if_expr=to_if_expr(expr);
-      if_expr.cond() = rename<level>(std::move(if_expr.cond()), ns);
-      rename_address<level>(if_expr.true_case(), ns);
-      rename_address<level>(if_expr.false_case(), ns);
+      index_exprt result = index_expr;
+
+      if(renamed_array.has_value())
+      {
+        result.array() = std::move(*renamed_array);
+        result.type() = to_array_type(result.array().type()).subtype();
+      }
 
-      if_expr.type()=if_expr.true_case().type();
+      if(renamed_index.has_value())
+        result.index() = std::move(*renamed_index);
+
+      return std::move(result);
     }
-    else if(expr.id()==ID_member)
+    else
+      return {};
+  }
+  else if(expr.id() == ID_if)
+  {
+    // the condition is not an address
+    const if_exprt &if_expr = to_if_expr(expr);
+    auto renamed_cond = rename_expr<level>(if_expr.cond(), ns);
+    auto renamed_true = rename_address<level>(if_expr.true_case(), ns);
+    auto renamed_false = rename_address<level>(if_expr.false_case(), ns);
+
+    if(
+      renamed_cond.has_value() || renamed_true.has_value() ||
+      renamed_false.has_value())
     {
-      member_exprt &member_expr=to_member_expr(expr);
+      if_exprt result = if_expr;
 
-      rename_address<level>(member_expr.struct_op(), ns);
+      if(renamed_cond.has_value())
+        result.cond() = std::move(*renamed_cond);
 
-      // type might not have been renamed in case of nesting of
-      // structs and pointers/arrays
-      if(
-        member_expr.struct_op().type().id() != ID_struct_tag &&
-        member_expr.struct_op().type().id() != ID_union_tag)
+      if(renamed_true.has_value())
       {
-        const struct_union_typet &su_type=
-          to_struct_union_type(member_expr.struct_op().type());
-        const struct_union_typet::componentt &comp=
-          su_type.get_component(member_expr.get_component_name());
-        PRECONDITION(comp.is_not_nil());
-        expr.type()=comp.type();
+        result.true_case() = std::move(*renamed_true);
+        result.type() = result.true_case().type();
       }
-      else
-        rename<level>(expr.type(), irep_idt(), ns);
+
+      if(renamed_false.has_value())
+        result.false_case() = std::move(*renamed_false);
+
+      return std::move(result);
     }
     else
-    {
-      // this could go wrong, but we would have to re-typecheck ...
-      rename<level>(expr.type(), irep_idt(), ns);
+      return {};
```

Comment:
The function is becoming quite big and it may be good to extract some static function for instance to make an optional `if_exprt` from an original expression and three optionals, but it doesn't need to be in this pull request.